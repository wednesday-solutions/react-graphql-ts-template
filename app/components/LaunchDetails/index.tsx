import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/react';
import { If, T, For } from '@components';
import isEmpty from 'lodash-es/isEmpty';
import { colors, media } from '@app/themes';
import { LaunchDetails as LaunchDetailsType } from '@app/containers/LaunchDetails/types';
import { Card, Skeleton } from '@mui/material';
import placeholderImage from '@images/undraw_to_the_stars_re_wq2x.svg';

const LaunchDetailsCard = styled(Card)`
  && {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    color: ${(props) => props.color};

    margin: 1.5rem;
    @media (min-width: ${media.tablet}) {
      flex-direction: row;
      gap: 1rem;
    }
    background-color: ${colors.cardBg};
  }
`;

const CustomImage = styled.img`
  && {
    width: 100%;
    height: auto;
    @media (min-width: ${media.tablet}) {
      width: 50%;
      max-height: 698px;
      object-fit: cover;
    }
  }
`;

const DetailsCard = styled.div`
  && {
    display: flex;
    flex-direction: column;
    justify-content: center;
    gap: 1rem;
    margin: 3rem 0;
    @media (min-width: ${media.tablet}) {
      width: 50%;
    }
  }
`;

const RocketBox = styled.div`
  && {
    margin-left: 1rem;
    display: flex;
    flex-direction: column;
    gap: 1rem;
  }
`;

const CustomT = styled(T)`
  && {
    display: block;
    font-weight: thin;
  }
`;

const launchLabelStyles = () => css`
  text-transform: uppercase;
  display: block;
  color: ${colors.primary};
  font-size: smaller;
  font-weight: bold;
`;

const LaunchLabel = styled.b`
  ${launchLabelStyles()}
`;

const LaunchLabelT = styled(T)`
  ${launchLabelStyles()}
`;

const ShipContainer = styled.div`
  margin-left: 1rem;
  display: grid;
  align-items: center;
  row-gap: 1rem;
`;

function labelRenderer(chunks: string) {
  return <LaunchLabel>{chunks}</LaunchLabel>;
}

interface LaunchDetailsProps extends LaunchDetailsType {
  loading: boolean;
}

const renderSkeleton = () => {
  return (
    <>
      <Skeleton data-testid="skeleton" animation="wave" variant="text" height={40} />
      <Skeleton data-testid="skeleton" animation="wave" variant="text" height={40} />
      <Skeleton data-testid="skeleton" animation="wave" variant="text" height={40} />
    </>
  );
};

function LaunchDetails({ missionName, links, details, rocket, ships, loading }: LaunchDetailsProps) {
  return (
    <LaunchDetailsCard data-testid="launch-details" variant="outlined">
      <If condition={!loading} otherwise={renderSkeleton()}>
        <If condition={!isEmpty(links?.flickrImages)} otherwise={<CustomImage src={placeholderImage} />}>
          <CustomImage src={links?.flickrImages![0]} />
        </If>
      </If>
      <If condition={!loading} otherwise={renderSkeleton()}>
        <DetailsCard>
          <If condition={!isEmpty(missionName)}>
            <CustomT marginBottom={0.5} data-testid="mission-name" type="heading" text={missionName} />
          </If>

          <If condition={!isEmpty(details)}>
            <CustomT data-testid="details" type="standard" id="details" values={{ details, b: labelRenderer }} />
          </If>
          <If condition={!isEmpty(rocket)}>
            <LaunchLabelT type="smallBoldText" id="rocket" />
            <RocketBox>
              <If condition={!isEmpty(rocket?.rocketName)}>
                <CustomT
                  data-testid="rocket-name"
                  type="standard"
                  id="name_label"
                  values={{ name: rocket?.rocketName, b: labelRenderer }}
                />
              </If>
              <If condition={!isEmpty(rocket?.rocketType)}>
                <CustomT
                  data-testid="rocket-type"
                  type="standard"
                  id="type_label"
                  values={{ type: rocket?.rocketType, b: labelRenderer }}
                />
              </If>
            </RocketBox>
          </If>
          <If condition={!isEmpty(ships)}>
            <LaunchLabelT type="smallBoldText" id="ships" />
            <ShipContainer>
              <For
                noParent
                of={ships}
                renderItem={(ship) => (
                  <>
                    <If condition={!isEmpty(ship.name)}>
                      <CustomT data-testid="ship-name" id="name_label" values={{ name: ship.name, b: labelRenderer }} />
                    </If>
                    <If condition={!isEmpty(ship.type)}>
                      <CustomT data-testid="ship-type" id="type_label" values={{ type: ship.type, b: labelRenderer }} />
                    </If>
                  </>
                )}
              />
            </ShipContainer>
          </If>
        </DetailsCard>
      </If>
    </LaunchDetailsCard>
  );
}

LaunchDetails.propTypes = {
  id: PropTypes.string,
  loading: PropTypes.bool,
  missionName: PropTypes.string,
  launchDateUtc: PropTypes.string,
  links: PropTypes.shape({
    flickrImages: PropTypes.arrayOf(PropTypes.string)
  }),
  details: PropTypes.string,
  rocket: PropTypes.shape({
    rocketName: PropTypes.string,
    rocketType: PropTypes.string
  }),
  ships: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      type: PropTypes.string
    })
  )
};

export default LaunchDetails;
