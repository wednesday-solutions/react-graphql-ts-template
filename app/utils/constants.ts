export const HEADER_HEIGHT = '7rem';
export const MIN_SIDEBAR_WIDTH = '4.5rem';
export const MOBILE_DRAWER_BREAKPOINT = 450;

export const SCREEN_BREAK_POINTS = {
  mobile: 320,
  tablet: 768,
  desktop: 992,
  largeDesktop: 1400
};
