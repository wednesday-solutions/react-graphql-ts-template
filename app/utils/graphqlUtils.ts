import ApolloClient, { DocumentNode, InMemoryCache } from 'apollo-boost';
import camelCase from 'lodash-es/camelCase';
import { mapKeysDeep } from '.';

export const client = new ApolloClient({
  uri: 'https://spacex-production.up.railway.app',
  cache: new InMemoryCache()
});

export interface GqlQueryReponse<Data> {
  data?: Data;
  error?: any;
  ok: boolean;
}

export const getQueryResponse = <Data, Variables>(
  query: DocumentNode,
  variables?: Variables
): Promise<GqlQueryReponse<Data>> => {
  return client
    .query<Data, Variables>({ query, variables })
    .then((res) => {
      if (res.errors) {
        return { error: res.errors, ok: false };
      } else {
        return { data: mapKeysDeep(res.data, camelCase), ok: true };
      }
    })
    .catch((err) => {
      return { error: err, ok: false };
    });
};
