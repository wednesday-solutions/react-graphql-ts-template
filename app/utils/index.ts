import find from 'lodash-es/find';
import get from 'lodash-es/get';
import history from './history';
import routeConstants from './routeConstants';

export const getCurrentRouteDetails = (location: Partial<Location>) => {
  if (!get(location, 'pathname')) {
    return null;
  }
  const route = find(
    Object.keys(routeConstants),
    (key) => routeConstants[key].route === location.pathname || `${routeConstants[key].route}/` === location.pathname
  );
  if (route) {
    return routeConstants[route];
  }
  return null;
};
export const mapKeysDeep = <T>(obj: T, fn: (key: string) => string): T =>
  Array.isArray(obj)
    ? obj.map((val) => mapKeysDeep(val, fn))
    : typeof obj === 'object'
    ? Object.keys(obj!).reduce((acc, current) => {
        const key = fn(current);
        const val = obj![current as keyof T];
        acc[key] = val !== null && typeof val === 'object' ? mapKeysDeep(val, fn) : val;
        return acc;
      }, {} as any)
    : obj;

export const isLocal = () => {
  try {
    if (process.env.IS_LOCAL) {
      const local = JSON.parse(process.env.IS_LOCAL);
      return typeof local === 'boolean' && local;
    }
  } catch {
    // continue regardless of error
  }
  return false;
};

interface SetQueryParamOptions {
  param: string;
  value?: string | number;
  deleteParam?: boolean;
  historyOp?: 'push' | 'replace';
}

export const setQueryParam = ({ param, value, deleteParam, historyOp = 'push' }: SetQueryParamOptions) => {
  const urlParams = new URLSearchParams(history.location.search);
  if (deleteParam) {
    urlParams.delete(param);
  } else {
    urlParams.set(param, String(value));
  }
  if (typeof history[historyOp] === 'function') {
    history[historyOp]({ search: urlParams.toString() });
  } else {
    throw new Error('Invalid history operation');
  }
};

export const prepare = (payload: any) => ({ payload });
