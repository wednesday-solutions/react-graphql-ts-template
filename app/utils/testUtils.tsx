import React from 'react';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from '@mui/material/styles';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import IntlGlobalProvider from '@components/IntlGlobalProvider';
import history from './history';
import { theme } from '@containers/App';

export const renderWithIntl = (children: React.ReactNode) =>
  render(
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <IntlGlobalProvider>{children}</IntlGlobalProvider>
    </IntlProvider>
  );

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider theme={theme}>
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});

export const getStyles = () => {};
