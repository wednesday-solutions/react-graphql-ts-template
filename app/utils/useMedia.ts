import useScreenType from 'react-screentype-hook';
import { SCREEN_BREAK_POINTS } from '@app/utils/constants';

export default function useMedia() {
  return useScreenType({
    mobile: SCREEN_BREAK_POINTS.mobile,
    tablet: SCREEN_BREAK_POINTS.tablet,
    desktop: SCREEN_BREAK_POINTS.desktop,
    largeDesktop: SCREEN_BREAK_POINTS.largeDesktop
  });
}
