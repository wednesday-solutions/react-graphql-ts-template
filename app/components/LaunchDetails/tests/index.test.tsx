import React from 'react';
import LaunchDetails from '../index';
import { renderProvider } from '@app/utils/testUtils';

describe('<LaundDetails> tests', () => {
  const launchDetails = {
    id: '1',
    loading: false,
    missionName: 'CRS-21',
    links: {
      flickrImages: ['https://farm9.staticflickr.com/8617/16789019815_f99a165dc5_o.jpg']
    },
    details: "SpaceX's 21st ISS resupply mission.",
    rocket: {
      rocketName: 'Falcon 9',
      rocketType: 'FT'
    },
    ships: [
      {
        name: 'Ship 1',
        type: 'Type 1'
      }
    ]
  };
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render the mission name if it is available', () => {
    const { getByTestId } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(getByTestId('mission-name')).toBeInTheDocument();
  });
  it('should render the ship name if it is available', () => {
    const { getByTestId } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(getByTestId('ship-name')).toBeInTheDocument();
  });
  it('should render the ship type if it is available', () => {
    const { getByTestId } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(getByTestId('ship-type')).toBeInTheDocument();
  });
  it('should render the rocket name if it is available', () => {
    const { getByTestId } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(getByTestId('rocket-name')).toBeInTheDocument();
  });
  it('should render the rocket type if it is available', () => {
    const { getByTestId } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(getByTestId('rocket-type')).toBeInTheDocument();
  });
  it('should render the skeleton  if loading is true', () => {
    launchDetails.loading = true;
    const { baseElement } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(baseElement.getElementsByClassName('skeleton')).toBeTruthy();
  });
});
