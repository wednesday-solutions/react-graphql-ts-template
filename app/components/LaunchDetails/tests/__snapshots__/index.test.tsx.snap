// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaundDetails> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered css-dnax8p e1uve3jx7"
      data-testid="launch-details"
    >
      <div
        class="ant-card-body"
      >
        <img
          class="css-1v87snz e1uve3jx6"
          src="https://farm9.staticflickr.com/8617/16789019815_f99a165dc5_o.jpg"
        />
        <div
          class="css-bzx5os e1uve3jx5"
        >
          <p
            class="e1uve3jx3 css-1cewy1q e17qkzd60"
            data-testid="mission-name"
          >
            CRS-21
          </p>
          <p
            class="e1uve3jx3 css-139w90b e17qkzd60"
            data-testid="details"
          >
            <b
              class="css-1mq3jm7 e1uve3jx2"
            >
              Details
            </b>
             SpaceX's 21st ISS resupply mission.
          </p>
          <p
            class="e1uve3jx1 css-rz7hlo e17qkzd60"
            data-testid="t"
          >
            Rocket 
          </p>
          <div
            class="css-1pbbfrk e1uve3jx4"
          >
            <p
              class="e1uve3jx3 css-139w90b e17qkzd60"
              data-testid="rocket-name"
            >
              <b
                class="css-1mq3jm7 e1uve3jx2"
              >
                Name
              </b>
               Falcon 9
            </p>
            <p
              class="e1uve3jx3 css-139w90b e17qkzd60"
              data-testid="rocket-type"
            >
              <b
                class="css-1mq3jm7 e1uve3jx2"
              >
                Type
              </b>
               FT
            </p>
          </div>
          <p
            class="e1uve3jx1 css-rz7hlo e17qkzd60"
            data-testid="t"
          >
            Ships 
          </p>
          <div
            class="css-wt1l7g e1uve3jx0"
          >
            <p
              class="e1uve3jx3 css-139w90b e17qkzd60"
              data-testid="ship-name"
            >
              <b
                class="css-1mq3jm7 e1uve3jx2"
              >
                Name
              </b>
               Ship 1
            </p>
            <p
              class="e1uve3jx3 css-139w90b e17qkzd60"
              data-testid="ship-type"
            >
              <b
                class="css-1mq3jm7 e1uve3jx2"
              >
                Type
              </b>
               Type 1
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
